### DISCUSSION TITLE
Visual Studio Code integration?

### DISCUSSION THREAD

**Discussion by ines** [explosion/spaCy#2969] [New Features & Project Ideas]
Just came across the PyTorch VSCode integration [on Twitter](https://twitter.com/PyTorch/status/1067147602418896896) 😍

![](https://pbs.twimg.com/media/Ds9FB4tVAAAKO-Y.jpg)

## Feature description

I started on a little spaCy snippets extension ages ago and never really quite finished it. But I always thought it'd be cool to have a spaCy extension with some helpers and maybe some deeper pipeline, data structures and model inspection tools. I haven't really worked with VSCode plugins (yet), but maybe someone from the community has an idea and/or experience? Would be cool to work on this together!

**Comment by julien-c** [comment]
Hi @ines! I've played with a spaCy-powered vscode extension myself, and can maybe lend a hand if needed :)

![vscode-embeddings_demo](https://user-images.githubusercontent.com/326577/49089200-023e0500-f229-11e8-8755-ba18ccc21750.gif)

**Comment by ines** [comment]
@julien-c Ohh, yes, I remember seeing this on Twitter!

One idea I had for a tooltip-style feature was showing the model meta when hovering a model name in `spacy.load`. Basically, here:

![screenshot 2018-11-27 at 17 28 09](https://user-images.githubusercontent.com/13643239/49096022-f4f53a80-f269-11e8-9c0f-f4fd84a37359.png)

Or maybe something in the style of [Import Cost](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost)?

![](https://file-wkbcnlcvbn.now.sh/import-cost.gif)

Assuming your Python path is set correctly in your editor, we don't usually need to load a model in order to get its `meta.json` – we can look for it in the the installed packages and `/spacy/data`. Alternatively, if this is too difficult to do fast enough / async, I guess it could also be a panel that opens to the side and shows the `meta.json`.

Other ideas:
* displaCy visualizers! Just not sure how this would look and how you'd activate/run it. I guess it's already possible via the [Jupyter integration](https://blogs.msdn.microsoft.com/pythonengineering/2018/11/08/data-science-with-python-in-visual-studio-code/).
* [Custom debugger](https://code.visualstudio.com/docs/extensions/example-debuggers) – sounds pretty involved, though, so that'd probably be a bigger project. But I think it'd be really cool if you could pause/inspect the `Doc` on every step of the pipeline (I always thought this could potentially be a nice spaCy plugin in general and pretty straightforward to implement by just wrapping `nlp.pipeline`).
* Auto-packaging for pipeline components! This is less spaCy-specific, but imagine you could highlight your pipeline component function/class, use a shortcut and get a Python package containing the component code, with entry points (available as of `v2.1.x`) and everything!
* Match pattern tester widget/panel, similar to our [matcher demo](https://explosion.ai/demos/matcher)

**Comment by svlandeg** [comment]
We now published the spaCy VSCode Extension, providings additional tooling and features for working with spaCy's config files 🥳 

https://marketplace.visualstudio.com/items?itemName=Explosion.spacy-extension
